packages/mimir/src/rules/parameter-properties.ts

```diff
@@ -0,0 +1,193 @@
+import {
+    getLineBreakStyle,
+    isCallExpression,
+    isExpressionStatement,
+    isClassDeclaration,
+    isClassExpression,
+    isConstructorDeclaration,
+    isParameterProperty,
+    isBinaryExpression,
+    isPropertyAccessExpression,
+    isIdentifier,
+    isPropertyDeclaration,
+    hasSideEffects,
+} from 'tsutils';
+
+export interface Options {
+    mode: 'when-possible' | 'never' | 'consistent';
+}
+
+@excludeDeclarationFiles
+export class Rule extends ConfigurableRule<Options> {
+    protected parseOptions(options: Options | null | undefined): Options {
+        return options || { mode: 'when-possible' };
+    }
+
+    public apply() {
+        const checkClass = (node: ts.Node): void => {
+            if (
+                (isClassDeclaration(node) || isClassExpression(node)) &&
+                node.members.filter(isConstructorDeclaration).length === 1
+            )
+                this.checkConstructorParameters(node.members.find(isConstructorDeclaration)!, node);
+
+            return ts.forEachChild(node, checkClass);
+        };
+        ts.forEachChild(this.context.sourceFile, checkClass);
+    }
+
+    private get failureString(): string {
+        switch (this.options.mode) {
+            case 'never':
+                return 'Parameter properties have been disallowed.';
+            case 'when-possible':
+                return 'Use parameter properties when possible.';
+            default:
+                return 'Either all constructor parameters must be parameter properties, or none.';
+        }
+    }
+
+    private checkConstructorParameters(
+        construct: ts.ConstructorDeclaration,
+        classNode: ts.ClassDeclaration | ts.ClassExpression,
+    ): void {
+        switch (this.options.mode) {
+            case 'never':
+                for (const param of construct.parameters.filter(isParameterProperty))
+                    this.addFailureAtNode(
+                        param,
+                        this.failureString,
+                        getFixerForDisallowedParameterProp(
+                            construct,
+                            param,
+                            getLineBreakStyle(this.context.sourceFile),
+                        ),
+                    );
+                break;
+            case 'when-possible':
+                for (const param of construct.parameters.filter(
+                    (p: ts.ParameterDeclaration) => !isParameterProperty(p) && canBeParameterProperty(p, construct),
+                ))
+                    this.addFailureAtNode(
+                        param,
+                        this.failureString,
+                        getFixerForLonghandProp(param, construct, classNode),
+                    );
+                break;
+            default:
+                if (construct.parameters.some(isParameterProperty) && !construct.parameters.every(isParameterProperty))
```

Comment:
this implementation of `consistent` doesn't detect if a parameter could be a parameter property.
IMO the option should detect if all parameters *could* be parameter properties and otherwise forbid them.